### PR TITLE
Functions for restoring from last or best checkpoint

### LINKED ISSUE #7927: [tune/rllib] Better Usability for restoring from last checkpoint
### Describe your feature request

Restoring from last checkpoint requires a bit of wrangling (`max(glob.glob)`). It would be nice if a user could pass in a directory, and we look for the most recent checkpoint and restore directly from that directory.

### PR DESCRIPTION
**Pull request by stefanbschneider**
## Why are these changes needed?

Utility functions in `tune.trainable` that simplify loading the last checkpoint from a given directory.
Before, users had to always specify a specific checkpoint path when using `restore()`. Now, they can simply pass the directory with all checkpoints to `restore_last_checkpoint()`.

Alternatively, I also added `restore_best_checkpoint()`, which loads the best checkpoint from the given directory according to a given metric and mode.

## Related issue number

Closes https://github.com/ray-project/ray/issues/7927

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by stefanbschneider** [comment]
Actually, it seems that my unit tests are failing: https://travis-ci.com/github/ray-project/ray/jobs/491777419#L817

I'll check the error and try to resolve it later today. My function to load the last checkpoint relies on the checkpoints being stored in subfolders containing the string `checkpoint` and on numbering of these checkpoints like `checkpoint_15`. Apparently this doesn't hold for the unit tests - not sure if it is generally a valid assumption?

File: python/ray/tune/trainable.py
Comment (krfricke) [inline python/ray/tune/trainable.py:416]: It seems like this tries to do something like `ray.tune.utils.trainable.TrainableUtils.find_checkpoint_dir()`.

Generally I think these functions should live in the TrainableUtil as well

File: python/ray/tune/trainable.py
Comment (krfricke) [inline python/ray/tune/trainable.py:448]: While passing the current training iteration as an index to `TrainableUtil.make_checkpoint_dir` it is not enforced and users can pass arbitrary strings here. Can we instead do something similar to the `get_best_checkpoint`, but just choose the checkpoint with the maximum number of iterations (Key stored in `ray.tune.result.TRAINING_ITERATION`)?

**Review by krfricke — CHANGES_REQUESTED** [review]
Generally this is great, thanks for the contribution! Let's move this to the `TrainableUtil` class and let's try to make it as general as possible - i.e. not relying on increasing checkpoint dir names.

File: python/ray/tune/trainable.py
Comment (stefanbschneider) [reply python/ray/tune/trainable.py:416]: Ok, thanks for the hint. I wasn't available of that module yet. I'll move the code there.

File: python/ray/tune/trainable.py
Comment (stefanbschneider) [reply python/ray/tune/trainable.py:448]: Sounds good, I'll give it a try! 

How should I identify checkpoint directories and files in general? Can I rely on them starting with `checkpoint` or is there a more generic approach? Is the `checkpoint` prefix stored in some constant somewhere that I could access?

Comment (krfricke) [reply python/ray/tune/trainable.py:448]: The `checkpoint` prefix is currently hardcoded in `TrainableUtil.make_checkpoint_dir()` - since we're moving these functions into the same class you can just hardcode it and don't have to read it from a constant. 

I think it's fine to rely on the prefix, but be sure to call `find_checkpoint_dir()` to ensure it is actually a checkpoint. We could potentially get rid of the filter for `checkpoint_`, but since that prefix is currently actually enforced we don't need to do that right now.

**Comment by stefanbschneider** [comment]
@krfricke I am struggling with setting up unit tests for testing my moved code. I think I need to run some training iterations with `tune.run()` to create proper checkpoints to later test if my new functions find these checkpoints successfully.
For running `tune.run()`, I thought of a similar setup as here: https://github.com/ray-project/ray/blob/master/python/ray/tune/tests/test_tune_restore.py#L43

Unfortunately, running `ray.init()` locally with my version of `ray` from `master` always crashes with `AttributeError: module 'ray.ray_constants' has no attribute 'REDIS_KEY_DASHBOARD'` which makes it impossible to test my unit tests and thus my code.
Any idea what could be the problem?

**Comment by stefanbschneider** [comment]
I'm currently struggling to get a proper dev install of Ray working. I'm following [these instructions](https://docs.ray.io/en/master/development.html#building-ray-python-only) on Windows 10 and Python 3.8 (which is what I use to run `ray` usually), but I keep getting errors simply when importing ray or at latest when running `ray.init()`.

I'm currently in the process of debugging it through trial and error and testing WSL as alternative. Once I have the setup running, I'll complete my unit tests and finish the PR.

**Comment by stefanbschneider** [comment]
I was able to get the dev install working on WSL, found a few bugs and fixed them. Now all tests for my code are passing locally and there are not formatting issues according to `format.sh`.

**Comment by stefanbschneider** [comment]
@krfricke As far as I can see, my tests are passing. The failing tests seem to be unrelated, eg, `python/ray/tests/test_dashboard.py::test_ray_start_default_port_conflict`.

I also added an example in the docs.
Is there anything else, I should do before the PR can be merged?

**Comment by krfricke** [comment]
Hi @stefanbschneider, yep, the failing test looks unrelated. I'll take a look later/early today!

**Comment by krfricke** [comment]
Hi @stefanbschneider, sorry for the delay.

After taking another look, I think we might want to do a couple of changes.

First, it seems like this is actually not something we'd want to move in the trainable util after all - the trainable util ist mostly used interally by the tune components, so we shouldn't expose these as a public API. I didn't initially grasp the exact use case for this, so apologies for misguiding you there.

Instead, we would probably want to move the functionality to the `Analysis` object. In fact, the `Analysis` object has a couple of abstractions that already do what we try to do. E.g. to find the best checkpoint, you could use:

```
        analysis = Analysis(self.logdir)
        best_checkpoint = analysis.get_best_checkpoint(
            analysis.get_best_logdir(metric, mode),
            metric, mode)
```

and to find the last checkpoint you could use:

```
        analysis = Analysis(self.logdir)
        last_checkpoint = analysis.get_best_checkpoint(
            analysis.get_best_logdir(metric, mode),
            "training_iteration", "max")
```

It seems instead of introducing these new utilities to TrainableUtil we'd probably want to introduce methods that wrap around this existing functionality. E.g. something like

```
class Analysis:
    # ...

    def get_best_trial_checkpoint(metric, mode, select="best")
        if select == "best":
            checkpoint_metric = metric
            checkpoint_mode = mode
        elif select == "last":
            checkpoint_metric = "training_iteration"
            checkpoint_mode = "max"
        else:
            raise ValueError(f"Select has to be one of [best, last]. Got: {select}")

        return self.get_best_checkpoint(self.get_best_logdir(metric, mode), checkpoint_metric, checkpoint_mode)
```
and `select` could be `best` (default) or `last`.

What do you think?

Also I think in any case we should have a test where multiple checkpoint are created, so that we can check it actually finds the most recent/best checkpoint. There should be some tests like that in `test_checkpoint_manager.py`

Again, sorry for initially giving a bad suggestion here. I should have looked at the use case more closely.

**Comment by stefanbschneider** [comment]
@krfricke Ah ok, no worries. I knew of `analysis.get_best_checkpoint()` (and used it to get the best checkpoint) but didn't think of it for also loading the last checkpoint.
I agree, then writing simple wrapper functions around that should be enough. No need to implement similar functionality twice.

I'll make the updates in the code and tests in the next days and will let you know once they're ready for review.

**Comment by stefanbschneider** [comment]
I thought about this again. I think loading the best checkpoint is already sufficiently easy with `Analysis` and doesn't need a wrapper. The main idea of the issue https://github.com/ray-project/ray/issues/7927 was to simplify loading the last checkpoint.
So I just made simple wrapper function for that. What do you think?

At least that would getting the last checkpoint easier for me and would also be straight forward. (I might not think of using a function `get_best_checkpoint` to get the last one).

File: .gitignore
Comment (krfricke) [inline .gitignore:168]: Can we undo this change? We probably don't want to match all files that might start with `venv`

File: python/ray/tune/analysis/experiment_analysis.py
Comment (krfricke) [inline python/ray/tune/analysis/experiment_analysis.py:293]: Hm, so if we run multiple trials, this will get the trial that ran for the most iterations, and will get the last checkpoint of that trial. 

Is this what we want? I thought we would want to get the last checkpoint of the _best_ trial.

In that case, shouldn't we just accept a metric and mode to fetch the best trial by that, and then return the last checkpoint of that trial?

**Review by krfricke — COMMENTED** [review]
Thanks for the changes. Just a general question on the API - but it looks fine otherwise!

File: python/ray/tune/analysis/experiment_analysis.py
Comment (stefanbschneider) [reply python/ray/tune/analysis/experiment_analysis.py:293]: Hm, good question. I'm not sure. I usually only had a single trial anyways and wanted to load the last checkpoint as easily as possible (few required arguments).

I posted some related questions in the issue as well since someone was just asking about it: https://github.com/ray-project/ray/issues/7927#issuecomment-809573299 
I'm happy to change the function to whatever we need/want. Or to implement two (or three) wrapper functions if that helps.

Comment (stefanbschneider) [reply python/ray/tune/analysis/experiment_analysis.py:299]: @krfricke I just updated the function to optionally accept a metric and mode like you suggested and then load the best trial according to these.
By making metric and mode optional, it's still possible to simply get the last checkpoint of a single trial without having to specify any arguments.

The function itself is now just a very simple wrapper of `get_best_checkpoint()`, but I think it's value is in being a simple, straight-forward way of getting the last checkpoint for users that are not familiar with all functions/API.

What do you think?

Comment (krfricke) [reply python/ray/tune/analysis/experiment_analysis.py:299]: Great! Just one last thing - I think the call in line 301 should always use `metric=training_iteration` and `mode=max`, right? Because then it will always fetch the last checkpoint, but select the best trial according to the passed metric and mode.

File: python/ray/tune/analysis/experiment_analysis.py
Comment (krfricke) [inline python/ray/tune/analysis/experiment_analysis.py:301]:
```suggestion
        return self.get_best_checkpoint(trial, "training_iteration", "max")
```

Be sure to run the linter script again if you accept this suggestions

File: python/ray/tune/analysis/experiment_analysis.py
Comment (krfricke) [reply python/ray/tune/analysis/experiment_analysis.py:299]: (I assume this is what we want since the function is called `get_last_checkpoint`)

Comment (stefanbschneider) [reply python/ray/tune/analysis/experiment_analysis.py:299]: Oh, of course. Thanks for noticing!

**Comment by stefanbschneider** [comment]
@krfricke Thanks for checking the PR again. I merged your suggested change and checked the corresponding unit tests and linter - both passed. So I think the PR should be good to go? :)

**Comment by krfricke** [comment]
Actually it seems the linter does not pass on CI - can you check that again? Otherwise it looks great and I'm happy to merge it after that!

Thanks again for addressing everything so quickly, and generally for your contribution!

**Review by krfricke — APPROVED** [review]
Thanks again!

**Pull request merged**
